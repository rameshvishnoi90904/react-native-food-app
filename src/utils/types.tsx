export type FoodItem = {
    id: number,
    label: string,
    description: string,
    displayOrder: number,
    unitPriceFractional: number,
    currency: string,
    imageUrl: string,
    itemStock: {
        quantityLeft: number
    }
}

export type InnerSectionList = {
    sectionId: string,
    sectionDescription: string,
    disabledReason: string,
    list: FoodItem[],
    subSection: SectionItem[]
}

export type SectionItem = {
    id: number,
    label: string,
    description: string,
    displayOrder: number,
    disabled: Boolean,
    disabledReason: string|null,
    data: InnerSectionList[],
    outerSectionTitle: string
}


export type NavigationScreenPropsType = {
    Home: undefined;
    Detail: {item: FoodItem, disabledSection: boolean};
};

export interface CartState {
    [key: number]: CartItem;
}

export type CartItem = {
    foodId: number,
    quantity: number
}

export type dispatchActionType = {
    type: string;
    foodId: number;
}
