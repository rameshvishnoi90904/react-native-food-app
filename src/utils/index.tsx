import { FoodItem, SectionItem } from "./types";
/**
 * 
 * @param item section item
 * @param sectionTitle outer nested section title
 * @returns section item with outer nested section title
 */
const formatInnerSection = (item: any, sectionTitle: string) => {
    const foodItems = [...item.items];
    foodItems.sort((a: FoodItem, b: FoodItem) => (a.displayOrder - b.displayOrder));

    const toReturn: SectionItem = {
        id: item.id, 
        label: item.label,
        outerSectionTitle: sectionTitle,
        description: item.description,
        disabled: item.disabled,
        disabledReason: item.disabledReason,
        displayOrder: item.displayOrder,
        data: [{list: item.items, sectionId: item.id, sectionDescription: item.description, disabledReason: item.disabled ? item.disabledReason : "", subSection: []}],
    };
    return toReturn;
}
/**
 * 
 * @param item section data from api
 * @returns section item compatible with react native SectionList component
 */
export const formatSection = (item: any) : SectionItem => {
    const foodItems = [...item.items];
    const subSections = [...item.subSections];

    foodItems.sort((a: FoodItem, b: FoodItem) => (a.displayOrder - b.displayOrder));
    subSections.sort((a: SectionItem, b: SectionItem) => (a.displayOrder - b.displayOrder));
    
    const formattedSubSection: SectionItem[] = subSections.map((sItem) => formatInnerSection(sItem, item.label))

    const toReturn = {
        id: item.id, 
        label: item.label,
        description: item.description,
        disabled: item.disabled,
        outerSectionTitle: "",
        disabledReason: item.disabledReason,
        displayOrder: item.displayOrder,
        data: [{list: item.items, sectionId: item.id, sectionDescription: item.description, disabledReason: item.disabled ? item.disabledReason : "", subSection: formattedSubSection}],
    };
    return toReturn;
}