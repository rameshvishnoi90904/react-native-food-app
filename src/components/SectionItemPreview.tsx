import React from "react"
import { View, Text, StyleSheet } from "react-native"
import { SectionItem } from "../utils/types";
const SectionItemPreview = ({section}:{section:SectionItem}) => {
    return (
        <View style={styles.item}>
            <Text style={styles.header}>{section.label}</Text>
            {
                section.outerSectionTitle ?
                <Text style={styles.subHeader}>({section.outerSectionTitle})</Text>
                :
                null
            }
        </View>
    )
}

const styles = StyleSheet.create({
    item: {
        backgroundColor: '#fff',
        paddingVertical: 10,
        // flexDirection: 'row'
    },
    header: {
        fontSize: 32,
        // backgroundColor: '#fff',
    },
    subHeader: {

    }
})
export default SectionItemPreview;