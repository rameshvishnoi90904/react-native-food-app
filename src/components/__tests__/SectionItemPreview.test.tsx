import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import {it, jest, expect} from '@jest/globals';
import SectionItemPreview from '../SectionItemPreview';
import { SectionItem } from '../../utils/types';

it('renders FoodItemPreview correctly without outer section', () => {
    const mockSectionItem : SectionItem = {
        id: 16,
        label: "Burgers",
        description: "Watching your carbs? Go bun-free and replace our wholemeal buns with a lettuce wrap!",
        displayOrder: 0,
        disabled: true,
        disabledReason: "Only available from 1st of May",
        data: [],
        outerSectionTitle: ""
    }
    const tree = render(<SectionItemPreview section={mockSectionItem}/>).toJSON();
    expect(tree).toMatchSnapshot();
})

it('renders FoodItemPreview correctly with outer section', () => {
    const mockSectionItem : SectionItem = {
        id: 16,
        label: "Burgers",
        description: "Watching your carbs? Go bun-free and replace our wholemeal buns with a lettuce wrap!",
        displayOrder: 0,
        disabled: true,
        disabledReason: "Only available from 1st of May",
        data: [],
        outerSectionTitle: "This is Outer Section Title"
    }
    const tree = render(<SectionItemPreview section={mockSectionItem}/>).toJSON();
    expect(tree).toMatchSnapshot();
})



