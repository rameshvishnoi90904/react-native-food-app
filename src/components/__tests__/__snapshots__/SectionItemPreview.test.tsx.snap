// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FoodItemPreview correctly with outer section 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 32,
      }
    }
  >
    Burgers
  </Text>
  <Text
    style={{}}
  >
    (
    This is Outer Section Title
    )
  </Text>
</View>
`;

exports[`renders FoodItemPreview correctly without outer section 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 32,
      }
    }
  >
    Burgers
  </Text>
</View>
`;
